src/ucs/datastruct/ptr_array.h

```diff
@@ -121,12 +145,230 @@ void *ucs_ptr_array_replace(ucs_ptr_array_t *ptr_array, unsigned index, void *ne
+ *
+ * @param [in]  locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in]  value             Pointer to insert. Must be 8-byte aligned.
+ *
+ * @return The index to which the value was inserted.
+ *
+ * Complexity: Amortized O(1)
+ *
+ * @note The array will grow if needed.
+ */
+unsigned ucs_ptr_array_locked_insert(ucs_ptr_array_locked_t *locked_ptr_array,
+                                     void *value);
+
+
+/**
+ * Remove a pointer from the locked array.
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index to remove from.
+ *
+ * Complexity: O(1)
+ */
+void ucs_ptr_array_locked_remove(ucs_ptr_array_locked_t *locked_ptr_array,
+                                 unsigned index);
+
+
+/**
+ * Replace pointer in the locked array
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ * @param [in] index             Index of slot
+ * @param [in] new_val           Value to put into slot given by index
+ *
+ * @return Old value of the slot
+ *
+ * Complexity: O(1)
+ */
+void *ucs_ptr_array_locked_replace(ucs_ptr_array_locked_t *locked_ptr_array,
+	                           unsigned index, void *new_val);
+
+
+/** Acquire the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_acquire_lock(_locked_ptr_array) \
+    ucs_recursive_spin_lock(&(_locked_ptr_array)->lock)
+
+
+/** Release the ptr_array lock
+ *
+ * @param [in] locked_ptr_array  Pointer to a locked ptr array.
+ */
+#define ucs_ptr_array_locked_release_lock(_locked_ptr_array) \
+    ucs_recursive_spin_unlock(&(_locked_ptr_array)->lock)
+
+
+/**
+ * Retrieves a value from the locked array.
+ *
+ * The user is required to lock using ucs_ptr_array_locked_acquire_lock()
+ * and ucs_ptr_array_locked_release_lock() functions.
+ *
+ * @param [in]  locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  index              Index to retrieve the value from.
+ * @param [out] var                Filled with the value.
+ *
+ * @return Whether the value is present and valid.
+ *
+ * Complexity: O(1)
+ */
+static inline int UCS_F_MAYBE_UNUSED
+ucs_ptr_array_locked_lookup(ucs_ptr_array_locked_t *locked_ptr_array,
+                            unsigned index,
+                            void **var)
+{
+    int present;
+
+    ucs_ptr_array_locked_acquire_lock(locked_ptr_array);
+    present = ucs_ptr_array_lookup(&(locked_ptr_array)->super, index, *var);
+    ucs_ptr_array_locked_release_lock(locked_ptr_array);
+
+    return present;
+}
+
+
+/**
+ * If foreach locked ptr_array is finalized, releases lock.
+ *
+ * @param [in]  _locked_ptr_array   Pointer to a locked ptr array.
+ * @param [in]  is_continue_loop    For() loop end condition.
+ *
+ * @return is_continue_loop for the for() loop end condition.
+ *
+ * Complexity: O(1)
+ */
+static inline int
+ucx_ptr_array_locked_foreach_finalize(ucs_ptr_array_locked_t *_locked_ptr_array,
```

Comment:
is this function intended for internal use only?
should we add `__` prefix to the function?

Reply:
Fixed.